python/cudf/cudf/core/indexing.py

```diff
@@ -38,6 +38,28 @@ def indices_from_labels(obj, labels):
     return lhs.join(rhs).sort_values("__")["_"]
 
 
+def get_label_range_or_mask(index, start, stop, step):
+    if (
+        not (start is None and stop is None)
+        and type(index) is cudf.core.index.DatetimeIndex
+        and index.is_monotonic is False
```

Comment:
I believe that the handling for monotonic `DatetimeIndex` and non-monotonic `DatetimeIndex` can be the same, i.e., we don't have to check for `index.is_monotonic is False` here.

That being said, maybe for the monotonic case it's worth going with the `find_label_range` approach if it's faster.

Reply:
As we will be just slicing the input for monotonic rather than creating complete new frame, I felt it might be memory critical, slicing would be faster than two binary operation with a gather.